clang/lib/AST/TypePrinter.cpp

```diff
@@ -1635,6 +1635,17 @@ void TypePrinter::printElaboratedBefore(const ElaboratedType *T,
     if (T->getKeyword() != ElaboratedTypeKeyword::None)
       OS << " ";
     NestedNameSpecifier *Qualifier = T->getQualifier();
+    if (!Policy.SuppressTagKeyword && Policy.SuppressScope &&
+        !Policy.SuppressUnwrittenScope) {
+      StringRef prefix = T->isClassType()       ? "class "
+                         : T->isStructureType() ? "struct "
+                         : T->isUnionType()     ? "union "
+                                                : "";
+      OS << prefix;
+      Policy.SuppressTagKeyword = true;
+      Policy.SuppressScope = false;
```

Comment:
Don't we want to reset these once we're done with the call to `printBefore()`?

Reply:
Yes, I think so.

Reply:
Hmmm, your changes don't look quite right to me. The `return printBefore(...);` is going to skip right over the restoration.

I think you wanted something more like:
```
      bool OldTag = Policy.SuppressTagKeyword, OldScope = Policy.SuppressScope;
      Policy.SuppressTagKeyword = true;
      Policy.SuppressScope = false;
      printBefore(T->getNamedType(), OS);
      Policy.SuppressTagKeyword = OldTag;
      Policy.SuppressScope = OldScope;
      return;
```
Before making the changes, I think you should also add test coverage to the unit test where these changes will fix the test.

Reply:
Not sure what you mean here "where these changes will fix the test"? will fix which test? The unit test you are referring to is `DeclPrinterTest`? this should be `TestCXXRecordDecl12`.

Reply:
Hmmm! Can't seem to find a test case that fails. At each step of  the printing  `ElaboratedTypePolicyRAII PolicyRAII(Policy);` gets called and that re-initializes the values of` Policy.SuppressTagKeyword` and  `Policy.SuppressScope` to true!

Reply:
You need to reset the values here, or use a separate `PrintingPolicy` object so that you're not changing the value for the default printing policy used.